src/coreclr/src/binder/assemblybinder.cpp

```diff
@@ -624,26 +608,29 @@ namespace BINDER_SPACE
         IF_FAIL_GO(hr);
         if (pContextEntry != NULL)
         {
-#if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
-            if (skipVersionCompatibilityCheck)
+            if (!skipVersionCompatibilityCheck)
             {
-                // Skip RefDef matching if we have been asked to.
+                // Can't give higher version than already bound
+                bool isCompatible = IsCompatibleAssemblyVersion(pAssemblyName, pContextEntry->GetAssemblyName());
+                hr = isCompatible ? S_OK : FUSION_E_APP_DOMAIN_LOCKED;
+                pBindResult->SetAttemptResult(hr, pContextEntry);
+
+                // TPA binder returns FUSION_E_REF_DEF_MISMATCH for incompatible version
+                if (hr == FUSION_E_APP_DOMAIN_LOCKED && isTpaListProvided)
```

Comment:
Anyone know why we do this (return a different HResult for TPA binder vs not)?
I didn't figure it out reading through the code. The comment said `For our new binding models, use a more descriptive error code than APP_DOMAIN_LOCKED for bind failures`, which didn't really clarify it for me either (especially since the error message for APP_DOMAIN_LOCKED is the one that actually mentions version?).